onnxruntime/core/graph/graph.cc

```diff
@@ -4019,131 +4019,178 @@ Node& Graph::FuseSubGraph(const IndexedSubGraph& sub_graph,
+    ONNX_NAMESPACE::NodeProto subgraph_node_proto{};
+    subgraph_node->ToProto(subgraph_node_proto);
+
+    ORT_RETURN_IF_ERROR(AddConstantProtoAsInitializer(subgraph_node_proto, model_path));
+  }
+
+  CopyInitializers(graph_to_inline, unique_id);
+  return Status::OK();
+}
+
+Status Graph::FunctionToGraph(const ONNX_NAMESPACE::FunctionProto& func_to_inline, const Path& model_path, Graph& dest_graph) {
+  auto to_node_arg = [this](const std::string& name) {
+    return &this->GetOrCreateNodeArg(name, nullptr);
+  };
+
+  // Process constant nodes first and create NodeArg for these as they become initializers
+  // It is important for the initializers to have NodeArg created, first they are needed
+  // if the initializer is unused and removed, second if the node depends on the initializer,
+  // we can have Type attached to it.
+  InlinedVector<const NodeProto*> non_constant_nodes;
+  non_constant_nodes.reserve(func_to_inline.node_size());
+  for (const auto& inlined_node : func_to_inline.node()) {
+    if (inlined_node.op_type() == kConstant) {
+      // Copy constant nodes _value to name_to_initial_tensor_
+      ORT_RETURN_IF_ERROR(dest_graph.AddConstantProtoAsInitializer(inlined_node, model_path));
+    } else {
+      non_constant_nodes.push_back(&inlined_node);
     }
+  }
 
-    for (auto& init : subgraph.name_to_initial_tensor_) {
-      const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
-      *tensor = *init.second;
-      tensor->set_name(tensor->name() + uniq_identifier);
-      auto insert_result = name_to_initial_tensor_.emplace(tensor->name(), tensor);
-      ORT_ENFORCE(insert_result.second, "Initializer name: ", tensor->name(), " in inlined subgraph: ",
-                  subgraph.Name(), " conflicts with graph initializer. Check Specializing code.");
-      if (GetNodeArg(tensor->name()) == nullptr) {
-        TypeProto t{TypeProtoFromTensorProto(*tensor)};
-        ORT_IGNORE_RETURN_VALUE(GetOrCreateNodeArg(tensor->name(), &t));
-      }
+  for (const auto* inlined_node : non_constant_nodes) {
+    InlinedVector<onnxruntime::NodeArg*> inputs;
+    InlinedVector<onnxruntime::NodeArg*> outputs;
+
+    for (const auto& tensor_name : inlined_node->input())
+      inputs.push_back(to_node_arg(tensor_name));
+
+    for (const auto& tensor_name : inlined_node->output())
+      outputs.push_back(to_node_arg(tensor_name));
+
+    onnxruntime::NodeAttributes new_attr_map;
+    new_attr_map.reserve(inlined_node->attribute_size());
+    for (const auto& node_attr : inlined_node->attribute()) {
+      new_attr_map.insert_or_assign(node_attr.name(), node_attr);
     }
+    dest_graph.AddNode(inlined_node->name(), inlined_node->op_type(),
+                       inlined_node->doc_string(), inputs, outputs, &new_attr_map, inlined_node->domain());
   }
 
-  RemoveNode(callnode.Index());
+  return Status::OK();
+}
+
+Status Graph::InlineFunction(Node& callnode) {
```

Comment:
nit: not clear what the 'call' in 'callnode' means. The arg is the function-based node being inlined isn't it?

Reply:
I do find it useful. It is meant to distinguish the caller (the node that represents a call to a function) from the callee (which is the called function). May not be a big difference here, but elsewhere we deal with nodes in the called function also.

Reply:
I am just preserving the current semantics.